libs/basic_execution/include/hpx/basic_execution/receiver.hpp

```diff
@@ -0,0 +1,365 @@
+                "invocable");
+            return HPX_BASIC_EXECUTION_RECEIVER_SET_DONE_IMPL_EXPRESSION;
+        }
+#undef HPX_BASIC_EXECUTION_RECEIVER_SET_DONE_IMPL_EXPRESSION
+#undef HPX_BASIC_EXECUTION_RECEIVER_SET_DONE_EXPRESSION
+
+#define HPX_BASIC_EXECUTION_RECEIVER_SET_DONE_EXPRESSION                       \
+    tag_invoke_impl(IsTagInvocable{}, std::forward<Receiver>(rcv))
+        template <typename Receiver,
+            typename IsTagInvocable = hpx::functional::is_nothrow_tag_invocable<
+                hpx::basic_execution::set_done_t, Receiver>>
+        constexpr HPX_FORCEINLINE auto operator()(Receiver&& rcv) const noexcept
+            -> decltype(HPX_BASIC_EXECUTION_RECEIVER_SET_DONE_EXPRESSION)
+        {
+            return HPX_BASIC_EXECUTION_RECEIVER_SET_DONE_EXPRESSION;
+        }
+#undef HPX_BASIC_EXECUTION_RECEIVER_SET_DONE_EXPRESSION
+    } set_done;
+
+    namespace detail {
+        template <bool ConstructionRequirements, typename T, typename E>
+        struct is_receiver_impl;
+
+        template <typename T, typename E>
+        struct is_receiver_impl<false, T, E> : std::false_type
+        {
+        };
+
+        template <typename T, typename E>
+        struct is_receiver_impl<true, T, E>
+          : std::integral_constant<bool,
+                hpx::traits::is_callable<hpx::basic_execution::set_done_t(
+                    typename std::decay<T>::type&&)>::value &&
+                    hpx::traits::is_callable<hpx::basic_execution::set_error_t(
+                        typename std::decay<T>::type&&, E)>::value>
+        {
+        };
+    }    // namespace detail
+
+    template <typename T, typename E>
+    struct is_receiver
+      : detail::is_receiver_impl<
+            std::is_move_constructible<typename std::decay<T>::type>::value &&
+                std::is_constructible<typename std::decay<T>::type, T>::value,
+            T, E>
+    {
+    };
+
+    template <typename T, typename E = std::exception_ptr>
+    constexpr bool is_receiver_v = is_receiver<T, E>::value;
+
+    namespace detail {
+        template <bool IsReceiver, typename T, typename... As>
+        struct is_receiver_of_impl;
+
+        template <typename T, typename... As>
+        struct is_receiver_of_impl<false, T, As...> : std::false_type
+        {
+        };
+
+        template <typename T, typename... As>
+        struct is_receiver_of_impl<true, T, As...>
+          : std::integral_constant<bool,
+                hpx::traits::is_callable<hpx::basic_execution::set_value_t(
+                    typename std::decay<T>::type&&, As...)>::value>
+        {
+        };
+    }    // namespace detail
+
+    template <typename T, typename... As>
+    struct is_receiver_of
+      : detail::is_receiver_of_impl<is_receiver_v<T>, T, As...>
+    {
+    };
+
+    template <typename T, typename... As>
+    constexpr bool is_receiver_of_v = is_receiver_of<T, As...>::value;
```

Comment:
Should these trait types go into `namespace hpx::traits`?

Reply:
I guess they should, to stay consistent with other parts of the library.